### DISCUSSION TITLE
Typing Subscriptions output and allowing input in subscriptions

### DISCUSSION THREAD

**Discussion by cabello** [trpc/trpc#6279] [Q&A]
Hello there,

I am following the subscriptions guide https://trpc.io/docs/server/subscriptions and checking the example code https://github.com/trpc/examples-next-sse-chat I am using the `httpSubscriptionLink` with React Query.

From looking at other issues, these versions might be relevant:
* `"typescript": "^5.7.2",`
* `"@trpc/server": "^11.0.0-rc.648",`

I also got the latest `zAsyncIterable` from the repo (https://github.com/trpc/trpc/blob/4b7757d9f426ab4eaa2c13153c2b3c5c5557eef5/packages/tests/server/zAsyncIterable.ts)

A simple random number worked as expected:

```
import { tracked } from "@trpc/server"
import { procedureBuilder } from "../init.js"

export const randomNumber = procedureBuilder
  .subscription(async function* () {
    while (true) {
      const next = Math.random()
      yield tracked(String("fake-id"), { draw: next })
      await new Promise(resolve => setTimeout(resolve, 1000))
    }
  })
```

However the data was not typed on the client side yet:

```
trpc.randomNumber.useSubscription(undefined, {
    // enabled: true,
    onData: data => {
      console.log(data)
    },
    onError: error => {
      console.error(error)
    },
  })
```

Once I added:

```
.output(
    zAsyncIterable({
      yield: z.object({ draw: z.number() }),
      tracked: true,
    }),
  )
```

Then Typescript would show me a valid looking type but the `console.log` and attempting to use that type would just explode because the raw type was a json list with 3 items:

```
{
    "json": [
        "fake-id",
        {
            "draw": 0.3428868238871339
        },
        null
    ]
}
```

So although the editor would let me do `data.id` or `data.data.draw`, the raw value was this list, missing the cast somewhere.

Any ideas what I am doing wrong, could this be an interference with `superjson`?

**Comment by cabello** [comment]
I just noticed with the current setup that by **not** using `.output(zAsyncIterable` my types on the client side are working as expected, they are an object and not the json array and I can use the attributes as expected.

**Reply by KATT** [reply]
I'm not at a laptop so can't test or tell you if you're doing something wrong, but I think we have tests for this. If we don't, feel free to fork and make a failing test in a PR and I'll make it work

**Reply by KATT** [reply]
Or just post an issue with a reproduction

**Comment by cabello** [comment]
As long as I don't try the official docs and return a typed async generator from my iterable, trpc has the correct type on the client side.
